### PR TITLE
BUG: RollingGroupby ignored as_index=False

### LINKED ISSUE #39433: BUG: groupby.rolling: Originial index is not being preserved when using date_part of DatetimeIndex and as_index key word seems to have no effect
- [ ] I have checked that this issue has not already been reported.

- [ x] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python
import pandas as pd

data = [
    ["A", "2018-01-01", 100],
    ["A", "2018-01-02", 200],
    ["B", "2018-01-01", 150],
    ["B", "2018-01-02", 250],
]
df = pd.DataFrame(data, columns=["id", "date", "num"])
df["date"] = pd.to_datetime(df["date"])
df.set_index(["date"], inplace=True)


df_res1 = df.groupby([df.id, df.index.weekday]).rolling(window=2, min_periods=1).mean()
df_res2 = df.groupby([df.id, df.index.weekday], as_index=False).rolling(window=2, min_periods=1).mean()
df_res3 = df.groupby([df.id]).rolling(window=2, min_periods=1).mean()
df_res4 = df.groupby([df.id], as_index=False).rolling(window=2, min_periods=1).mean()
pd.show_versions()
print(df_res1.head())
print(df_res1.index.names)
print(df_res2.index.names)
print(df_res3.index.names)
print(df_res4.index.names)


if __name__ == '__main__':
    ...
```

#### Problem description

Comparing pandas 1.0.1 and 1.2.1 yields:
1. When using a date part like df.index.weekday to group by and using some rolling window function like .mean() the original datetime index is not part of the resulting frames index anymore.
2. The group by keyword argumen as_index seems to have no effect in that situation anymore

#### Expected Output
['id', 'date', 'date']
[None, 'date']
['id', 'date']
[None, 'date']

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 9d598a5e1eee26df95b3910e3f2934890d062caa
python           : 3.8.4.final.0
python-bits      : 64
OS               : Linux
OS-release       : 4.15.0-45-generic
Version          : #48-Ubuntu SMP Tue Jan 29 16:28:13 UTC 2019
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8
pandas           : 1.2.1
numpy            : 1.19.1

</details>

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #39433
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

### CODE REVIEW

**Comment by phofl** [comment]
Does this fix #31007? Haven't looked closely but seen this in the past

**Comment by mroeschke** [comment]
@phofl unfortunately doesnt look like it

```
In [1]: import pandas as pd
   ...:
   ...: data = {
   ...:     'groupby_col': ['A', 'A', 'A', 'A', 'A', 'B', 'B', 'B', 'B', 'B', ],
   ...:     'agg_col': [1, 1, 0, 1, 0, 0, 0, 0, 1, 0],
   ...: }
   ...: df = pd.DataFrame(data)
   ...: df.groupby(['groupby_col'], as_index=False).rolling(4).agg({'agg_col': 'mean'})
Out[1]:
               agg_col
groupby_col
A           0      NaN
            1      NaN
            2      NaN
            3     0.75
            4     0.50
B           5      NaN
            6      NaN
            7      NaN
            8     0.25
            9     0.25
```

`agg` still goes partially through groupby code I think to calculate the index result.

**Comment by phofl** [comment]
Pitty, thanks for checking

File: pandas/core/window/rolling.py [hunk 1]
```diff
@@ -619,6 +621,8 @@ def _apply(
         )
 
         result.index = result_index
+        if not self._as_index:
+            result = result.reset_index(level=list(range(len(groupby_keys))))
```
Comment (rhshadrach) [inline pandas/core/window/rolling.py:625]: What happens here when the groupby is on an explicit list, e.g. in your test use `groupby(["A", "A", "B", "B"])` instead. What is groupby_keys in this case?

Comment (mroeschke) [reply pandas/core/window/rolling.py:625]: Here's the result

```
In [4]: df.groupby(["A", "A", "B", "B"], as_index=False).rolling(window=2, min_periods=1).mean()
> /Users/matthewroeschke/pandas-mroeschke/pandas/core/window/rolling.py(580)_apply()
-> result_index_names = groupby_keys + grouped_index_name
(Pdb) groupby_keys
[None]
(Pdb) c
Out[4]:
                num
  date
A 2018-01-01  100.0
  2018-01-02  150.0
B 2018-01-01  150.0
  2018-01-02  200.0

In [5]: df.groupby(["A", "A", "B", "B"], as_index=False).mean()
Out[5]:
     num
0  150.0
1  200.0
```

it appears in either case the grouping list isn't mapped as a column on the dataframe

File: pandas/tests/window/test_groupby.py [hunk 2]
```diff
@@ -732,6 +732,42 @@ def test_groupby_level(self):
         )
         tm.assert_series_equal(result, expected)
 
+    def test_as_index_false(self):
```
Comment (jreback) [inline pandas/tests/window/test_groupby.py:735]: can you add the multi-key groupby here as well (parameterize if you can)

```
df_res2 = df.groupby([df.id, df.index.weekday], as_index=False).rolling(window=2, min_periods=1).mean()
df_res3 = df.groupby([df.id]).rolling(window=2, min_periods=1).mean()
df_res4 = df.groupby([df.id], as_index=False).rolling(window=2, min_periods=1).mean()
```
e.g. 2 & 4 (we likley have 1 & 3 covered, but wouldn't object to those included as well)

Comment (mroeschke) [reply pandas/tests/window/test_groupby.py:735]: Hmm all these tested (except the `as_index=True` cases which are tested everywhere else). So you just want it parameterized?

Comment (jreback) [reply pandas/tests/window/test_groupby.py:735]: if they are elsewhere then don't (just the cases that is covering in this PR are fine). parameterize if you can.

**Comment by jreback** [comment]
thanks @mroeschke

**Pull request merged**